tests/python/pants_test/reporting/test_reporting_integration.py

```diff
@@ -265,6 +268,70 @@ def test_zipkin_reporter_for_v2_engine(self):
         v2_span_name_part, trace
         ))
 
+  def run_mock_server(self, bin_name, arguments):
+    args = ["../../../build-support/bin/native/cargo.sh", "run", "--release", "--bin", bin_name]
+    if len(arguments) > 0:
+      args = args + ["--"] + arguments
+    process = subprocess.Popen(args, cwd="src/rust/engine", stdin = subprocess.PIPE, stdout = subprocess.PIPE, stderr=subprocess.STDOUT, universal_newlines = True)
+    port = None
+    while port is None:
+      match = re.search("localhost:([0-9]+)", process.stdout.readline())
+      if match is not None:
+        port = int(match.group(1))
+      time.sleep(0.1)
+    return (process, port)
+
+  def run_cas_server(self):
+    return self.run_mock_server("local_cas", []);
+
+  def run_execution_server(self, request_hash, request_size):
+    return self.run_mock_server("local_execution_server", ["--request_hash", request_hash, "--request_size", request_size])
+
+  def test_zipkin_reporter_for_remote_execution_with_v2_engine(self):
+    (cas_process, cas_port) = self.run_cas_server()
+    # Hardcode the hash and size of the request we know our call to cloc will trigger
+    # If this changes for some reason, the local mock execution server will print out
+    # the digest of the request that was expected and this can be manually updated.
+    request_hash = "b22d7feb653a82f63961c9335922c523a42d588139f493ad78442d29c8e31928"
+    request_size = "142"
+    (execution_process, execution_port) = self.run_execution_server(request_hash, request_size)
+    ZipkinHandler = zipkin_handler()
+    random_store_location = secrets.token_hex(8)
+    with http_server(ZipkinHandler) as port:
+      endpoint = "http://localhost:{}".format(port)
+      command = [
+        'cloc',
+        'src/python/pants:version',
+        '--reporting-zipkin-trace-v2',
+        '--reporting-zipkin-endpoint={}'.format(endpoint),
+        '--remote-execution',
+        '--remote-execution-server=localhost:{}'.format(execution_port),
+        '--remote-store-server=localhost:{}'.format(cas_port),
+        '--cache-cloc-ignore',
+        '--local-store-dir=/tmp/cloc/store/{}'.format(random_store_location)
+      ]
+      pants_run = self.run_pants(command)
+      self.assert_success(pants_run)
+
+      # Terminate both the cas and the local execution server process by sending them an
+      # end of line character
+      cas_process.communicate("\n")
```

Comment:
This will leak the processes if the pants run failed (and so we stopped running the function when we hit the assertion failure on line 314). Consider a context manager, along the lines of:

```python
with self.run_cas_server() as cas_port, self.run_execution_server() as execution_port:
  do stuff
```

You can see how this works if you look at some examples in src/python/pants/util/contextutil.py - basically, you do your initialisation, you `yield` any value you want to return out (in this case the port), and after the `yield` you do your cleanup (in this case, sending the newline)